From 78d11d5050c3f5b6c202e0ff4503b38f38492702 Mon Sep 17 00:00:00 2001
From: Khem Raj <raj.khem@gmail.com>
Date: Tue, 7 Mar 2023 21:21:57 -0800
Subject: [PATCH] PCI: imx6: Provide a clock to the device for i.MX8MQ

When the internal PLL is configured as PCIe REF_CLK, we also have to
output a clock via CLK2_P/N pin to the connector/device to provide it.
Configure 100 MHz clock as its output.

Signed-off-by: Ryosuke Saito <rsaito@redhat.com>
Signed-off-by: Khem Raj <raj.khem@gmail.com>
[rebased onto the enable_ext_refclk flag added in v7.2]
Signed-off-by: Rudi Heitbaum <rudi@heitbaum.com>
---
 drivers/pci/controller/dwc/pci-imx6.c | 35 +++++++++++++++++++++++++++
 1 file changed, 35 insertions(+)

diff --git a/drivers/pci/controller/dwc/pci-imx6.c b/drivers/pci/controller/dwc/pci-imx6.c
index 5248ed0..af2d7fd 100644
--- a/drivers/pci/controller/dwc/pci-imx6.c
+++ b/drivers/pci/controller/dwc/pci-imx6.c
@@ -43,6 +43,11 @@
 #define IMX8MQ_GPR_PCIE_CLK_REQ_OVERRIDE	BIT(11)
 #define IMX8MQ_GPR_PCIE_VREG_BYPASS		BIT(12)
 #define IMX8MQ_GPR12_PCIE2_CTRL_DEVICE_TYPE	GENMASK(11, 8)
+#define IMX8MQ_ANA_PLLOUT_MONITOR_CFG_REG	0x74
+#define IMX8MQ_ANA_PLLOUT_MONITOR_CLK_SEL_MASK	GENMASK(3, 0)
+#define IMX8MQ_ANA_PLLOUT_MONITOR_CKE		BIT(4)
+#define IMX8MQ_ANA_SCCG_PLLOUT_DIV_CFG_REG	0x7C
+#define IMX8MQ_ANA_SCCG_SYSPLLL1_DIV_MASK	GENMASK(2, 0)
 
 #define IMX95_PCIE_PHY_GEN_CTRL			0x0
 #define IMX95_PCIE_REF_USE_PAD			BIT(17)
@@ -455,6 +460,36 @@ static int imx8mq_pcie_init_phy(struct imx_pcie *imx_pcie)
 			   imx_pcie_grp_offset(imx_pcie),
 			   IMX8MQ_GPR_PCIE_REF_USE_PAD,
 			   ext ? IMX8MQ_GPR_PCIE_REF_USE_PAD : 0);
+
+	if (!ext) {
+		/*
+		 * The internal PLL is used as REF_CLK, so the clock also has
+		 * to be provided to the device via the CLK2_P/N pins.
+		 */
+		struct regmap *anatop =
+			syscon_regmap_lookup_by_compatible("fsl,imx8mq-anatop");
+
+		if (IS_ERR(anatop)) {
+			dev_err(imx_pcie->pci->dev,
+				"Couldn't configure the internal PLL as REF clock\n");
+		} else {
+			/* Select SYSTEM_PLL1_CLK as the clock source */
+			regmap_update_bits(anatop, IMX8MQ_ANA_PLLOUT_MONITOR_CFG_REG,
+					   IMX8MQ_ANA_PLLOUT_MONITOR_CLK_SEL_MASK, 0xb);
+
+			/*
+			 * SYSTEM_PLL1_CLK is 800 MHz, so divide by 8 to
+			 * generate the 100 MHz output.
+			 */
+			regmap_update_bits(anatop, IMX8MQ_ANA_SCCG_PLLOUT_DIV_CFG_REG,
+					   IMX8MQ_ANA_SCCG_SYSPLLL1_DIV_MASK, 0x7);
+
+			/* Enable CLK2_P/N clock to provide it to the device */
+			regmap_update_bits(anatop, IMX8MQ_ANA_PLLOUT_MONITOR_CFG_REG,
+					   IMX8MQ_ANA_PLLOUT_MONITOR_CKE,
+					   IMX8MQ_ANA_PLLOUT_MONITOR_CKE);
+		}
+	}
 	/*
 	 * Per the datasheet, the PCIE_VPH is suggested to be 1.8V.  If the
 	 * PCIE_VPH is supplied by 3.3V, the VREG_BYPASS should be cleared
-- 
2.47.1

